### PR TITLE
[WIP][multi-lora] rewrite: tinker-compatible multi-LoRA backend

### PR DESCRIPTION
**Pull request by yushengsu-thu**
Clean and refactor now (https://github.com/radixark/miles/pull/2365): split into
- [x] backend, trainer
- [x] rollout（e2e）
- [x] rollout (tinker)
- [ ] ray (TrainGroup, ServingRef, publication,  E2E gate)
- [ ] tinker (tinker_sdk/)

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -165,10 +165,8 @@ RUN pip install "git+https://github.com/NVIDIA-NeMo/Emerging-Optimizers.git@v0.3
 RUN TMS_CUDA_MAJOR=$(python3 -c "import torch; print(torch.version.cuda.split('.')[0])") \
     pip install git+https://github.com/fzyzcjy/torch_memory_saver.git@f05a8754daf68238d54e4cf31cb3ba866684bbaf --no-cache-dir --force-reinstall
 RUN pip install "nvidia-modelopt[torch]>=0.37.0" --no-build-isolation
-# radixark/Megatron-Bridge#24 (TE 2.17 grouped-linear contract), merged to @bridge.
-# Pinned by SHA, not branch: buildkit caches this layer on the instruction text, so a
-# branch that moves leaves the old revision baked into the image with nothing to show it.
-RUN pip install git+https://github.com/radixark/Megatron-Bridge.git@7f0fb3456f8ffe47599b5fd167b454605d85f932 --no-deps --no-build-isolation
+# Multi-LoRA uses the grouped-GEMM fixes from radixark/Megatron-Bridge's bridge branch.
+RUN pip install git+https://github.com/radixark/Megatron-Bridge.git@bb61fcd0b61f8acd0ef0f8b38b2240968c94c37b --no-deps --no-build-isolation
```
Comment (yushengsu-thu) [inline docker/Dockerfile]: ping to `bridge` branch

Comment (yushengsu-thu) [reply docker/Dockerfile]: Updated the install ref to `@bridge` in 78481c8aa. Verified the current branch head (`bb61fcd0`) on H200 with the dense and MoE grouped-GEMM focused tests (5 passed).

Comment (yushengsu-thu) [reply docker/Dockerfile]: done

File: miles/utils/multi_lora.py [hunk 2]
```diff
@@ -31,6 +34,30 @@ class EmptyBatchTimeoutError(RuntimeError):
     """No trainable groups arrived before empty-wait timeout."""
 
 
+@dataclass(frozen=True)
+class AdapterIdentity:
```
Comment (yushengsu-thu) [inline miles/utils/multi_lora.py:42]: maybe change name to `LoraIdentity`

Comment (yushengsu-thu) [reply miles/utils/multi_lora.py:42]: In all multi-lora related files (including e2e or tinker) class, function, var with `adapter` prefix or postfix need to be changed to `lora`

File: miles/backends/training_utils/loss_hub/losses.py [hunk 3]
```diff
@@ -517,3 +517,30 @@ def get_loss_function(args: Namespace) -> LossFunction:
...
         case _:
             raise ValueError(f"Unknown loss type: {args.loss_type}")
+
+
+# Request-scoped loss names (tinker wire vocabulary) mapped onto existing kernels; other names are rejected upstream.
+REQUEST_LOSS_FUNCTIONS: dict[str, LossFunction] = {
+    "cross_entropy": sft_loss_function,
+    "importance_sampling": policy_loss_function,
+    "ppo": policy_loss_function,
+}
```
Comment (yushengsu-thu) [inline miles/backends/training_utils/loss_hub/losses.py]: move this to line 16

Comment (yushengsu-thu) [reply miles/backends/training_utils/loss_hub/losses.py]: Moved: the mapping now sits at the top of losses.py as a plain name-to-loss_type table (kernel objects are defined further down, so the resolver goes through get_loss_function at call time).

Comment (yushengsu-thu) [reply miles/backends/training_utils/loss_hub/losses.py]: done

**Comment by yushengsu-thu** [comment]
Backend, Trainer:
- ServiceClient.create_lora_training_client() → multi_lora_create_model()
- TrainingClient.forward() → multi_lora_forward()
- TrainingClient.forward_backward() → multi_lora_forward_backward()
- TrainingClient.optim_step() → multi_lora_optim_step()
- TrainingClient.save_state() → multi_lora_save_state()
- TrainingClient.load_state() → multi_lora_load_state(with_optimizer=False)
- TrainingClient.load_state_with_optimizer() → multi_lora_load_state(with_optimizer=True)
- TrainingClient.save_weights_for_sampler() → multi_lora_export_adapter() — only the export candidate is currently implemented; SGLang publication and commit are not yet implemented.

File: miles/backends/megatron_utils/multi_lora_utils.py [hunk 4]
```diff
@@ -431,43 +431,14 @@ def save_multi_lora_checkpoints(
...
-
 def install_adapter_slot(
     model, slot: int, *, rank: int, alpha: float, seed: int | None = None, state_dict: dict | None = None
 ) -> int:
-    """Policy-free slot install on this rank's model shard; returns the number of loaded tensors (0 = fresh init)."""
-    from megatron.bridge.peft.multi_lora_layers import clear_adapter_slot, init_adapter_slot, load_adapter
+    """Policy-free slot install on this rank's model shard; returns the number of loaded tensors (0 = fresh init).
+
+    Weight init is owned by Megatron-Bridge; a seed makes a fresh init reproducible (ignored when loading a state_dict).
+    """
```
Comment (yushengsu-thu) [inline miles/backends/megatron_utils/multi_lora_utils.py]: remove

Comment (yushengsu-thu) [reply miles/backends/megatron_utils/multi_lora_utils.py]: Removed; the docstring is back to one line.

Comment (yushengsu-thu) [reply miles/backends/megatron_utils/multi_lora_utils.py]: done

**Comment by yushengsu-thu** [comment]
Backend, Trainer:

1. refactor backend trainer to supported both multi-lora in non-interactive (e2e) and interactive (tinker) modes.
2. The interactive (tinker) mode require the following functions: (Tinker operation -> miles functions:)
- ServiceClient.create_lora_training_client() → multi_lora_create_model()
- TrainingClient.forward() → multi_lora_forward()
- TrainingClient.forward_backward() → multi_lora_forward_backward()
- TrainingClient.optim_step() → multi_lora_optim_step()
- TrainingClient.save_state() → multi_lora_save_state()
- TrainingClient.load_state() → multi_lora_load_state(with_optimizer=False)
- TrainingClient.load_state_with_optimizer() → multi_lora_load_state(with_optimizer=True)
- TrainingClient.save_weights_for_sampler() → multi_lora_export_adapter() — only the export candidate is currently implemented; SGLang publication and commit are not yet implemented.

File: miles/backends/megatron_utils/multi_lora_optimizer.py [hunk 5]
```diff
@@ -154,24 +161,40 @@ def zero_adapter_slot_grads(model, slot: int) -> None:
...
+        # Copy accumulated main_grads into the owned masters' grads, then apply the caller's scale to the sum.
         for child in children:
             child.prepare_grads()
-            for main_param in child.get_parameters():
-                if main_param.grad is not None:
-                    main_param.grad.mul_(1.0 / batch_size)
+            if scale != 1.0:
+                for main_param in child.get_parameters():
+                    if main_param.grad is not None:
+                        main_param.grad.mul_(scale)
```
Comment (yushengsu-thu) [inline miles/backends/megatron_utils/multi_lora_optimizer.py:197]: note: check if need any optimization in computing on later

**Comment by yushengsu-thu** [comment]
(Above) Backend, Trainer:

1. Refactor the backend and trainer to add interactive (Tinker) support to the existing non-interactive (E2E) multi-LoRA backend.
2. The interactive (tinker) operation/SDK require the following functions: (tinker operations SDK → miles multi-lora functions):
- ServiceClient.create_lora_training_client() → multi_lora_create_model()
- TrainingClient.forward() → multi_lora_forward()
- TrainingClient.forward_backward() → multi_lora_forward_backward()
- TrainingClient.optim_step() → multi_lora_optim_step()
- TrainingClient.save_state() → multi_lora_save_state()
- TrainingClient.load_state() → multi_lora_load_state(with_optimizer=False)
- TrainingClient.load_state_with_optimizer() → multi_lora_load_state(with_optimizer=True)
- TrainingClient.save_weights_for_sampler() → multi_lora_export_adapter() — only the export candidate is currently implemented; SGLang publication and commit are not yet implemented.

File: miles/rollout/sglang_rollout.py [hunk 6]
```diff
@@ -138,13 +138,51 @@ def submit_generate_tasks(self, samples: list[list[Sample]]) -> None:
...
+        url = f"http://{args.sglang_router_ip}:{args.sglang_router_port}/generate"
+    return await post(url, payload, headers=headers)
+
+
+def parse_output_token_logprobs(output: dict[str, Any]) -> tuple[list[int], list[float]]:
+    """Split raw /generate meta_info output_token_logprobs into response token ids and logprobs."""
+    if "output_token_logprobs" not in output["meta_info"]:
+        return [], []
+    pairs = output["meta_info"]["output_token_logprobs"]
+    return [item[1] for item in pairs], [item[0] for item in pairs]
```
Comment (yushengsu-thu) [inline miles/rollout/sglang_rollout.py]: might I write this function with original formate:
```python
    if "output_token_logprobs" in output["meta_info"]:
        new_response_tokens = [item[1] for item in output["meta_info"]["output_token_logprobs"]]
        new_response_log_probs = [item[0] for item in output["meta_info"]["output_token_logprobs"]]
    else:
        new_response_tokens, new_response_log_probs = [], []
```

Comment (yushengsu-thu) [reply miles/rollout/sglang_rollout.py]: Done — restored the original if/else format in `parse_output_token_logprobs` (pure move from the old `generate()` body). Amended into 5db098e16; equivalence suite re-run green.

**Comment by yushengsu-thu** [comment]
(Above) rollout:
1. abstract `async def call_sglang_generate_endpoint` and `def parse_output_token_logprobs` in `miles/rollout/sglang_rollout.py` for tinker re-using

**Comment by yushengsu-thu** [comment]
(Above) rollout (e2e):
1. abstract `async def call_sglang_generate_endpoint(...)` and `def parse_output_token_logprobs(...)` in `miles/rollout/sglang_rollout.py` for tinker re-using
2. Updated `async_rollout.py` and `data_source.py`: each LoRA must complete its full rollout before Trainer admission.
3. Preserved fully asynchronous multi-LoRA rollout and multi-slot batch coalescing in `async_rollout.py`.

**Comment by yushengsu-thu** [comment]
(Above) rollout (e2e):
1. abstract `async def call_sglang_generate_endpoint(...)` and `def parse_output_token_logprobs(...)` in `miles/rollout/sglang_rollout.py` for tinker re-using
2. Changed codes in `async_rollout.py` and `data_source.py`: each LoRA must complete its full rollout before Trainer admission.
3. Preserved fully asynchronous multi-LoRA rollout and multi-slot batch coalescing in `async_rollout.py`.

**Comment by yushengsu-thu** [comment]
(Above) Rollout (tinker): - most effort on miles/tinker

- Pin Tinker 0.26.1 and validate the runtime only when the Tinker frontend starts.
- Keep RolloutManager responsible only for SGLang fleet/router lifecycle; create no E2E or Trainer resources.
- Implement sampling HTTP, Futures, admission limits, and abort handling; call SGLang directly.
- Decode Datum/loss inputs while preserving row order, adapter identity, and request ownership.
- Validate the unmodified SDK through Trainer-ready input without executing training.

**Comment by yushengsu-thu** [comment]
Implement here: https://github.com/radixark/miles/pull/2846
